src/wayland/dmabuf/mod.rs

```diff
@@ -11,471 +11,470 @@
+        }
+    }
+
+    /// Creates a dmabuf global with the specified supported formats.
+    pub fn create_global<D, L>(
+        &mut self,
+        display: &mut Display<D>,
+        formats: Vec<Format>,
+        logger: L,
+    ) -> DmabufGlobal
+    where
+        D: GlobalDispatch<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1, GlobalData = DmabufGlobalData>
+            + BufferHandler
+            + DmabufHandler
+            + 'static,
+        L: Into<Option<::slog::Logger>>,
+    {
+        self.create_global_with_filter(display, formats, |_| true, logger)
+    }
+
+    /// Creates a dmabuf global with the specified supported formats.
+    ///
+    /// This function unlike [`DmabufState::create_global`] also allows you to specify a filter function to
+    /// determine which clients may see this global. This functionality may be used on multi-gpu systems in
+    /// order to make a client choose the correct gpu.
+    pub fn create_global_with_filter<D, F, L>(
+        &mut self,
+        display: &mut Display<D>,
+        formats: Vec<Format>,
+        filter: F,
+        logger: L,
+    ) -> DmabufGlobal
+    where
+        D: GlobalDispatch<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1, GlobalData = DmabufGlobalData>
+            + BufferHandler
+            + DmabufHandler
+            + 'static,
+        F: for<'c> FnMut(&'c Client) -> bool + Send + 'static,
+        L: Into<Option<::slog::Logger>>,
+    {
+        let id = next_global_id();
+        let logger = crate::slog_or_fallback(logger)
+            .new(slog::o!("smithay_module" => "wayland_dmabuf", "global" => id));
+        let formats = Arc::new(formats);
+        let data = DmabufGlobalData {
+            filter: Arc::new(Mutex::new(filter)),
+            formats,
+            id,
+            logger,
+        };
+
+        let global = display.create_global::<zwp_linux_dmabuf_v1::ZwpLinuxDmabufV1>(GLOBAL_VERSION, data);
+        self.globals.insert(id, global);
+
+        DmabufGlobal { id }
+    }
+
+    /// Disables a dmabuf global.
+    ///
+    /// This operation is permanent and there is no way to re-enable a global.
+    pub fn disable_global<D: 'static>(&mut self, display: &mut Display<D>, global: &DmabufGlobal) {
+        display.disable_global(self.globals.get(&global.id).unwrap().clone())
+    }
+
+    /// Destroys a dmabuf global.
+    ///
+    /// It is highly recommended you disable the global before destroying it and ensure all child objects have
+    /// been destroyed.
+    pub fn destroy_global<D: 'static>(&mut self, display: &mut Display<D>, global: DmabufGlobal) {
+        display.remove_global(self.globals.remove(&global.id).unwrap());
+        DMABUF_GLOBAL_IDS.lock().unwrap().remove(&global.id);
+    }
+}
+
+/// Data associated with a dmabuf global.
+#[allow(missing_debug_implementations)]
+pub struct DmabufGlobalData {
+    #[allow(clippy::type_complexity)]
+    filter: Arc<Mutex<dyn for<'c> FnMut(&'c Client) -> bool + Send>>,
```

Comment:
Why does this need to be in an `Arc`? It does not seem to be cloned anywhere.

Reply:
I believe it's for send or sync, I'll look at it again.